File: ObjectOrientedProgrammming/02-Stack-RPN-Solid/src/main/java/Stack/Stack.java
package Stack;
import java.util.EmptyStackException;

public class Stack {
    private int[] stacky;
    private int top = -1;

    // Construction site
    public Stack(){
        stacky = new int[11]; // Str stacky --> { null, null, ... , null } ; int stacky --> [I@adress ; int stacky = new int[11] --> [0 0 0 0 0 0 0 0 0 0 0 ...]
    }

    public Stack(int size){
        stacky = new int[size];
    }

    // private methods

    private void resize(){
        int[] Temp = new int[stacky.length*2];
        for (int i = 0; i < stacky.length; i++){  // System.arraycopy
            Temp[i] = stacky[i];
        }
        stacky = Temp;
    }

    // public methods

    public void push(int newEl){
        if (top < (stacky.length-1)){
            //System.out.println("top:" + top);
            //System.out.println("size:" + (size -1));
            top += 1;
            stacky[top] = newEl;
            //System.out.println("element: " + newEl + " on: " + top);
        }
        else{
            resize();
            //System.out.println("resized");
            push(newEl);
            //System.out.println("try again");
        }
    }

    public int pop() throws EmptyStackException {
        int result;
        if (top > -1){
            result = stacky[top];
            stacky[top] = 0; // default int is 0; Elements in primitive arrays can't be empty (null) so default int is inserted
            top -= 1;
        }
        else{
            throw new EmptyStackException();
        }
        return result;
    }

    public int peak(){ // Sneak Peak - get peak but don't pop it
        int result;
        if (top > -1){
            result = stacky[top];
        }
        else{
            throw new EmptyStackException();
        }
        return result;
    }
    public boolean isEmpty(){
        return top == -1;
    }



}

File: Data_Structures/01_Stack/src/Stack.py
class Stack:
    def __init__(self):
        self.structure = []

    def add(self, val):
        assert isinstance(val, (str, int, float)), "This element can't be in Stack\nU can add str, int or float"
        self.structure.append(val)

    def get(self):
        if self.is_empty():
            return 'Stack is empty'
        return self.structure.pop()

    def peak(self):
        if self.is_empty():
            return 'Stack is empty'
        return self.structure[-1]

    def __str__(self):
        return str(self.structure)

    def is_empty(self):
        if len(self.structure) == 0:
            return True
        else:
            return False
